ZIP-321: Payment Request URIs

nuttycom opened the pull request:
Closes https://github.com/zcash/zips/issues/347

nuttycom commented:
@ChileBob this may be of interest to you.

nuttycom commented:
@adityapk00 I'd like to draw your attention to this (standardization of zcash payment URIs) as well.

ChileBob commented:
A request, could a few fully formatted URIs be provided as examples?

pacu reviewed (commented):
+1 to @ChileBob's request on some samples.

daira commented on zip-0321.rst:
```diff
@@ -0,0 +1,168 @@
+References
+==========
+
+.. [#BIP0021] `BIP-0021 <https://en.bitcoin.it/wiki/BIP_0021>`_
+.. [#BIP0070] `BIP-0070 <https://en.bitcoin.it/wiki/BIP_0070>`_
+.. [#RFC3986] `RFC 3986 <https://tools.ietf.org/html/rfc3986#appendix-A>`_
```
```suggestion

.. [#RFC2119] `RFC 2119: Key words for use in RFCs to Indicate Requirement Levels <https://www.rfc-editor.org/rfc/rfc2119.html>`_
.. [#RFC5234] `RFC 5234: Augmented BNF for Syntax Specifications: ABNF <https://www.rfc-editor.org/rfc/rfc5234.html>`_
.. [#RFC3986] `RFC 3986: URI Generic Syntax, Appendix A. Collected ABNF for URI <https://www.rfc-editor.org/rfc/rfc3986.html#appendix-A>`_
.. [#bip-0021] `BIP 21: URI Scheme <https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki>`_
.. [#bip-0070] `BIP 70: Payment Protocol <https://github.com/bitcoin/bips/blob/master/bip-0070.mediawiki>`_
```

daira commented on zip-0321.rst:
```diff
@@ -0,0 +1,168 @@
+   Label for an address (e.g. name of receiver)
+
+address
+   Zcash address (shielded or transparent)
```
```suggestion
   Zcash address string (shielded or transparent)
```

daira commented on zip-0321.rst:
```diff
@@ -0,0 +1,168 @@
+   Zcash address (shielded or transparent)
+
+memo
+   Contents for the zcash shielded memo field. MUST not exceed 512 bytes.
```
```suggestion
   Contents for the Zcash shielded memo field. MUST not exceed 512 bytes.
```

daira commented on zip-0321.rst:
```diff
@@ -0,0 +1,168 @@
+
+Specification
+=============
+
```
```suggestion

The following syntax specification uses ABNF [#RFC5234]_.

```

daira commented on zip-0321.rst:
```diff
@@ -0,0 +1,168 @@
+the URI MUST contain an address parameter having that ``paramindex``.
+
+Transfer amount
+--------------------
```
```suggestion
---------------
```

daira commented on zip-0321.rst:
```diff
@@ -0,0 +1,168 @@
+message
+   Message that clients can display for the purpose of presenting descriptive
+   information about the transaction to the user.
+
```
Add an Examples section here. It should also include invalid examples with the reason(s) why they are invalid.

nuttycom replied:
I'll produce these from the reference implementation I'm working on.

daira reviewed (changes requested):
Suggested some more minor changes to the syntax, and addition of an Examples section.

daira reviewed (commented):
minor comments

nuttycom commented on zip-0321.rst:
```diff
@@ -0,0 +1,175 @@
+.. code-block:: EBNF
+
+  zcashurn        = "zcash:" ( zcashaddress [ "?" zcashparams ] / "?" zcashparams )
+  zcashaddress    = *base58
```
What is the parsing ambiguity? I'll need to know about it for the reference implementation I'm working on.

nuttycom commented on zip-0321.rst:
```diff
@@ -0,0 +1,175 @@
+
+If there are any non-address parameters having a given ``paramindex``, then 
+the URI MUST contain an address parameter having that ``paramindex``.
+
```
```suggestion

Productions of the form ``1*x`` indicate one or more successive instances of the production ``x``.
```

daira commented on zip-0321.rst:
```diff
@@ -0,0 +1,176 @@
+If there are any non-address parameters having a given ``paramindex``, then 
+the URI MUST contain an address parameter having that ``paramindex``.
+
+Productions of the form ``1*x`` indicate one or more successive instances of the production ``x``.
```
```suggestion
Productions of the form ``1*x`` indicate one or more successive instances of the
production ``x``.

```

daira commented on zip-0321.rst:
```diff
@@ -0,0 +1,176 @@
+
+References
+==========
+
```
```suggestion
```

nuttycom commented on zip-0321.rst:
```diff
@@ -0,0 +1,178 @@
+  addrparam       = "address" [ paramindex ] "=" zcashaddress
+  amountparam     = "amount"  [ paramindex ] "=" 1*DIGIT [ "." 1*DIGIT ]
+  labelparam      = "label"   [ paramindex ] "=" *qchar
+  memoparam       = "memo"    [ paramindex ] "=" *qchar
```
It occurs to me here that maybe *qchar isn't the correct type for the memo parameter; we want to be able to encode 512 arbitrary bytes, and from my sample generation URI-encoded unicode (and anything else apart from ASCII) looks like the dog's breakfast. Would it make sense here to standardize on base58-encoded bytes (since we use base58 encoding elsewhere)? cc/ @daira

daira replied:
We only use base58 for Sprout addresses. I think we should use base64.

nuttycom replied:
One potential issue with using base64 is that it specifies `=` for padding bytes, and would need to be hex-encoded. Base58 doesn't have this problem.

daira replied:
The `=` padding is not actually needed; you can unambiguously infer the input length from the encoded length even without padding.

nuttycom replied:
Does this make it more complicated to specify/implement though? IIRC most base64 libraries automatically add padding, so implementers would have to be cautious of this.

daira commented on zip-0321.rst:
```diff
@@ -0,0 +1,178 @@
+If an amount is provided, it MUST be specified in decimal ZEC. Amounts MUST NOT
+contain commas, and if a decimal fraction is present then a period (`.`) MUST be
+used as the separating character to separate the whole number from the decimal
+fraction. The decimal fraction MUST NOT be empty. E.g. ``amount=50.00`` or
+``amount=50`` is treated as 50 ZEC; ``amount=.5`` is treated as 0.5 ZEC; and
+``amount=50,000.00`` or ``amount=50,00`` or ``amount=50.`` are invalid.
```
```suggestion
fraction. Both the whole number and the decimal fraction MUST be nonempty.
E.g. ``amount=50.00`` or ``amount=50`` is treated as 50 ZEC; ``amount=0.5`` is
treated as 0.5 ZEC; and ``amount=50,000.00`` or ``amount=50,00`` or ``amount=50.``
or ``amount=.5`` are invalid.
```

nuttycom commented:
@ChileBob @pacu @adityapk00 I've added some sample URLs and addressed a number of changes that were requested by internal reviewers. Also, a reference implementation in Rust is available here: https://github.com/zcash/librustzcash/pull/294

Please let me know if you have additional questions or suggestions!

daira reviewed (approved):
ACK modulo changes.

pacu commented on zip-0321.rst:
```diff
@@ -6,5 +6,218 @@
+Transfer amount
+---------------
+
+If an amount is provided, it MUST be specified in decimal ZEC. Amounts MUST NOT
```
Wouldn't it be better to express the formatting in terms of Grouping/Thousand and decimal separators?

"The whole (integer) part must not contain grouping/thousands separators" or similar. People implementing this will find these terms in the number formatting classes of their language of choice.

nuttycom replied:
The `MUST NOT contain commas` bit here is actually redundant with the grammar, but the restriction is not limited to the whole part of the amount. Maybe "Amounts MUST NOT contain commas or other separators apart from the decimal point."?

pacu replied:
it seems pretty clear to me, but I googled it and it results that some locales use spaces as thousand separators 😱

daira commented on zip-0321.rst:
```diff
@@ -6,5 +6,231 @@
+Addresses, amounts, labels, and messages sharing the same ``paramindex`` are
+interpreted to be associated with the same payment for the purposes of payment
+construction. Parameters with an empty ``paramindex`` are associated with a
+payment to the address with empty ``paramindex``. Note that a ``paramindex`` of
+``.0`` is distinct from an empty ``paramindex``. A ``paramindex`` MUST NOT have
+leading zero(s).
```
```suggestion
payment to the address with empty ``paramindex``. An empty ``paramindex`` MUST
be treated as equivalent to a ``paramindex`` of ``.0``. A ``paramindex`` MUST NOT
have leading zero(s). There is no significance to the ordering of parameters, and
``paramindex`` values need not be sequential.
```

daira commented on zip-0321.rst:
```diff
@@ -6,5 +6,231 @@
+.. code-block:: 
+
+  zcash:?amount=1.234&amount=2.345&address=tmEZhbWHTpdKMw5it8YDspUXSMGQyFwovpU
+  zcash:?amount.0=1.234&amount.0=2.345&address.0=tmEZhbWHTpdKMw5it8YDspUXSMGQyFwovpU
```
```suggestion
  zcash:?amount.0=1.234&amount.0=2.345&address.0=tmEZhbWHTpdKMw5it8YDspUXSMGQyFwovpU
  zcash:?amount=1.234&amount.0=2.345&address.0=tmEZhbWHTpdKMw5it8YDspUXSMGQyFwovpU
```

daira commented on zip-0321.rst:
```diff
@@ -6,5 +6,233 @@
+
+Addresses, amounts, labels, and messages sharing the same ``paramindex`` are
+interpreted to be associated with the same payment for the purposes of payment
+construction. Parameters with an empty ``paramindex`` are associated with a
+payment to the address with empty ``paramindex``. An empty ``paramindex`` MUST
+be treated as equivalent to a ``paramindex`` of ``.0``. A ``paramindex`` MUST NOT
+have leading zero(s). There is no significance to the ordering of parameters, and
+``paramindex`` values need not be sequential.
```
```suggestion
construction. An empty ``paramindex`` MUST be treated as equivalent to a
``paramindex`` of ``.0``. A ``paramindex`` MUST NOT have leading zero(s).
There is no significance to the ordering of parameters, and ``paramindex``
values need not be sequential.
```

nuttycom commented on zip-0321.rst:
```diff
@@ -4,7 +4,234 @@
+Transfer amount
+---------------
+
+If an amount is provided, it MUST be specified in decimal ZEC. Amounts MUST NOT
+contain commas, and if a decimal fraction is present then a period (`.`) MUST be
+used as the separating character to separate the whole number from the decimal
+fraction. Both the whole number and the decimal fraction MUST be nonempty.
+Leading zeros in the whole number or trailing zeros in the decimal fraction
+are ignored. There MUST NOT be more than 8 digits in the decimal fraction.
+For example,
```
```suggestion
If an amount is provided, it MUST be specified in decimal ZEC. If a decimal fraction is
present then a period (`.`) MUST be used as the separating character to separate the whole
number from the decimal fraction, and both the whole number and the decimal fraction MUST
be nonempty. No other separators (such as commas for grouping or thousands) are permitted.
Leading zeros in the whole number or trailing zeros in the decimal fraction are ignored.
There MUST NOT be more than 8 digits in the decimal fraction.

For example,
If an amount is provided, it MUST be specified in decimal ZEC. If a decimal fraction is
present then a period (`.`) MUST be used as the separating character to separate the whole
number from the decimal fraction, and both the whole number and the decimal fraction MUST
be nonempty. No other separators (such as commas for grouping or thousands) are permitted.
Leading zeros in the whole number or trailing zeros in the decimal fraction are ignored.
There MUST NOT be more than 8 digits in the decimal fraction.

For example,
```

nuttycom replied:
@pacu ^^

daira commented on zip-0321.rst:
```diff
@@ -4,7 +4,238 @@
+
+  zcash:?address=tmEZhbWHTpdKMw5it8YDspUXSMGQyFwovpU&amount=1&amount.1=2&address.2=ztestsapling10yy2ex5dcqkclhc7z7yrnjq2z6feyjad56ptwlfgmy77dmaqqrl9gyhprdx59qgmsnyfska2kez
+
+Also invalid; this request is missing ``address.1=``
```
```suggestion
Also invalid; this request is missing ``address.1=``.
```

daira commented on zip-0321.rst:
```diff
@@ -4,7 +4,238 @@
+
+As this ZIP is written, several clients already implement a ``zcash:`` URI
+scheme similar to this one, however usually without the additional ``req-``
+prefix requirement. Thus, it is RECOMMENDED that additional variables prefixed
+with ``req-`` not be used in a mission-critical way until a grace period of 6
+months from the finalization of this ZIP has passed in order to allow client
+developers to release new versions, and users of old clients to upgrade.
```
I made a suggested change to this paragraph; I think it might have got lost.

daira replied:
```suggestion
prefix requirement or the facility to specify multiple payments using
``paramindex``. Thus, it is RECOMMENDED that these features not be used in
a mission-critical way until a grace period of 6 months from the finalization
of this ZIP has passed, in order to allow client developers to release new
versions, and users of old clients to upgrade.
```

daira commented on zip-0321.rst:
```diff
@@ -4,7 +4,241 @@
+   Zcash address string (shielded or transparent)
+
+memo
+   Contents for the Zcash shielded memo field, encoded as base64 without
```
```suggestion
   Contents for the Zcash shielded memo field, encoded as base64url without
```

daira commented on zip-0321.rst:
```diff
@@ -4,7 +4,241 @@
+
+  zcash:ztestsapling10yy2ex5dcqkclhc7z7yrnjq2z6feyjad56ptwlfgmy77dmaqqrl9gyhprdx59qgmsnyfska2kez?amount=1&memo=VGhpcyBpcyBhIHNpbXBsZSBtZW1vLg&message=Thank%20you%20for%20your%20purchase
+
+A valid payment request for a payment of 1 ZEC to a single shielded address, with a base64url-encoded memo and a message for display by the wallet.
```
```suggestion
A valid payment request for a payment of 1 ZEC to a single shielded address,
with a base64url-encoded memo and a message for display by the wallet.
```

daira commented on zip-0321.rst:
```diff
@@ -4,7 +4,241 @@
+
+  zcash:?address=tmEZhbWHTpdKMw5it8YDspUXSMGQyFwovpU&amount=123.456&memo=eyAia2V5IjogIlRoaXMgaXMgYSBKU09OLXN0cnVjdHVyZWQgbWVtby4iIH0&address.1=ztestsapling10yy2ex5dcqkclhc7z7yrnjq2z6feyjad56ptwlfgmy77dmaqqrl9gyhprdx59qgmsnyfska2kez&amount.1=0.789&memo.1=VGhpcyBpcyBhIHVuaWNvZGUgbWVtbyDinKjwn6aE8J-PhvCfjok
+
+A valid payment request with one transparent and one shielded recipient address, with encoded JSON and unicode memos.
```
```suggestion
A valid payment request with one transparent and one shielded recipient address,
with encoded JSON and Unicode memos.
```

nathan-at-least commented on zip-0321.rst:
```diff
@@ -4,7 +4,243 @@
+.. code-block:: EBNF
+
+  zcashurn        = "zcash:" ( zcashaddress [ "?" zcashparams ] / "?" zcashparams )
+  zcashaddress    = 1*base58
```
Does this line mean a zcash address begins with a `1` prefix? If so, that's incorrect.

nuttycom replied:
No, this means that the `zcashaddress` production is one or more base58 characters. See https://github.com/zcash/zips/pull/395/files#diff-a386636de36a003f2f6c54ee53696106R127. Also you'll note that this production has changed in a later amendment to the PR, and some more qualifications on addresses were added in https://github.com/zcash/zips/commit/8c4e0b5e147edaa9b14415a19e315b5e75368786#diff-a386636de36a003f2f6c54ee53696106R152

nathan-at-least reviewed (commented):
Clarification about `zcashaddress` grammar.
